File: api/v1/schemas/technical_chart.py
# -*- coding: utf-8 -*-
"""Technical chart API schemas (technical-v1)."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ChartLevel(BaseModel):
    price: float
    label: str
    source: str
    date: Optional[str] = None


class ChartExtreme(BaseModel):
    price: float
    date: str


class TechnicalChartItem(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: Optional[float] = None
    amount: Optional[float] = None
    change_percent: Optional[float] = None

    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    volume_ratio: Optional[float] = None
    volume_status: Optional[str] = None
    macd_dif: Optional[float] = None
    macd_dea: Optional[float] = None
    macd_bar: Optional[float] = None
    rsi6: Optional[float] = None
    rsi12: Optional[float] = None
    rsi24: Optional[float] = None
    boll_upper: Optional[float] = None
    boll_mid: Optional[float] = None
    boll_lower: Optional[float] = None
    boll_bandwidth: Optional[float] = None
    boll_position: Optional[float] = None
    kdj_k: Optional[float] = None
    kdj_d: Optional[float] = None
    kdj_j: Optional[float] = None
    cci: Optional[float] = None
    bias5: Optional[float] = None
    bias10: Optional[float] = None
    bias20: Optional[float] = None


class TechnicalChartSummary(BaseModel):
    latest_close: Optional[float] = None
    latest_change_percent: Optional[float] = None
    latest_volume_status: Optional[str] = None
    volume_ratio: Optional[float] = None
    support_levels: List[ChartLevel] = Field(default_factory=list)
    resistance_levels: List[ChartLevel] = Field(default_factory=list)
    recent_high: Optional[ChartExtreme] = None
    recent_low: Optional[ChartExtreme] = None


class TechnicalChartResponse(BaseModel):
    stock_code: str
    stock_name: Optional[str] = None
    period: str
    range_days: int
    calculation_version: str
    data_status: str
    data_source: Optional[str] = None
    updated_at: Optional[str] = None
    items: List[TechnicalChartItem] = Field(default_factory=list)
    summary: TechnicalChartSummary = Field(default_factory=TechnicalChartSummary)
    warnings: List[str] = Field(default_factory=list)

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "stock_code": "600519",
            "stock_name": "贵州茅台",
            "period": "daily",
            "range_days": 120,
            "calculation_version": "technical-v1",
            "data_status": "available",
            "data_source": "akshare",
            "updated_at": "2026-07-14T15:00:00+08:00",
            "items": [],
            "summary": {},
            "warnings": [],
        }
    })

File: apps/dsa-web/src/api/technicalChart.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export const TECHNICAL_CHART_INDICATOR_GROUPS = [
  'ma',
  'volume',
  'macd',
  'rsi',
  'boll',
  'kdj',
  'cci',
  'bias',
  'support_resistance',
] as const;

export type TechnicalChartIndicatorGroup = (typeof TECHNICAL_CHART_INDICATOR_GROUPS)[number];

/** Default to the complete technical-chart workspace on every viewport. */
export const TECHNICAL_CHART_DEFAULT_INDICATORS = [
  'ma',
  'boll',
  'volume',
  'macd',
  'rsi',
  'kdj',
  'cci',
  'bias',
  'support_resistance',
].join(',');

export const TECHNICAL_CHART_MOBILE_DEFAULT_INDICATORS = TECHNICAL_CHART_DEFAULT_INDICATORS;

export const TECHNICAL_CHART_ALLOWED_DAYS = [60, 120, 250] as const;
export type TechnicalChartDays = (typeof TECHNICAL_CHART_ALLOWED_DAYS)[number];
export const TECHNICAL_CHART_DEFAULT_DAYS: TechnicalChartDays = 60;

export const TECHNICAL_CHART_ALLOWED_PERIODS = ['daily', 'weekly', 'monthly'] as const;
export type TechnicalChartPeriod = (typeof TECHNICAL_CHART_ALLOWED_PERIODS)[number];
export const TECHNICAL_CHART_DEFAULT_PERIOD: TechnicalChartPeriod = 'daily';

export interface ChartLevel {
  price: number;
  label: string;
  source: string;
  date?: string | null;
}

export interface ChartExtreme {
  price: number;
  date: string;
}

export interface TechnicalChartItem {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number | null;
  amount?: number | null;
  changePercent?: number | null;
  ma5?: number | null;
  ma10?: number | null;
  ma20?: number | null;
  ma30?: number | null;
  ma60?: number | null;
  ma90?: number | null;
  ma120?: number | null;
  ma250?: number | null;
  volumeRatio?: number | null;
  volumeStatus?: string | null;
  macdDif?: number | null;
  macdDea?: number | null;
  macdBar?: number | null;
  rsi6?: number | null;
  rsi12?: number | null;
  rsi24?: number | null;
  bollUpper?: number | null;
  bollMid?: number | null;
  bollLower?: number | null;
  bollBandwidth?: number | null;
  bollPosition?: number | null;
  kdjK?: number | null;
  kdjD?: number | null;
  kdjJ?: number | null;
  cci?: number | null;
  bias5?: number | null;
  bias10?: number | null;
  bias20?: number | null;
}

export interface TechnicalChartSummary {
  latestClose?: number | null;
  latestChangePercent?: number | null;
  latestVolumeStatus?: string | null;
  volumeRatio?: number | null;
  supportLevels: ChartLevel[];
  resistanceLevels: ChartLevel[];
  recentHigh?: ChartExtreme | null;
  recentLow?: ChartExtreme | null;
}

export type TechnicalChartDataStatus = 'available' | 'partial' | 'empty' | 'source_unavailable';

export interface TechnicalChartResponse {
  stockCode: string;
  stockName?: string | null;
  period: string;
  rangeDays: number;
  calculationVersion: string;
  dataStatus: TechnicalChartDataStatus | string;
  dataSource?: string | null;
  updatedAt?: string | null;
  items: TechnicalChartItem[];
  summary: TechnicalChartSummary;
  warnings: string[];
}

export interface GetTechnicalChartParams {
  stockCode: string;
  period?: TechnicalChartPeriod;
  days?: TechnicalChartDays | number;
  indicators?: string;
  signal?: AbortSignal;
}

export function parseTechnicalChartDays(raw: string | null | undefined): TechnicalChartDays {
  const value = Number(raw);
  if (TECHNICAL_CHART_ALLOWED_DAYS.includes(value as TechnicalChartDays)) {
    return value as TechnicalChartDays;
  }
  return TECHNICAL_CHART_DEFAULT_DAYS;
}

export function parseTechnicalChartIndicators(raw: string | null | undefined): string {
  if (!raw || !raw.trim()) {
    return TECHNICAL_CHART_DEFAULT_INDICATORS;
  }
  const allowed = new Set<string>(TECHNICAL_CHART_INDICATOR_GROUPS);
  const parts = raw
    .split(',')
    .map((part) => part.trim().toLowerCase())
    .filter((part) => allowed.has(part));
  if (parts.length === 0) {
    return TECHNICAL_CHART_DEFAULT_INDICATORS;
  }
  // Deduplicate while preserving order.
  return Array.from(new Set(parts)).join(',');
}

export function parseTechnicalChartPeriod(raw?: string | null): TechnicalChartPeriod {
  const value = (raw || '').trim().toLowerCase();
  if (TECHNICAL_CHART_ALLOWED_PERIODS.includes(value as TechnicalChartPeriod)) {
    return value as TechnicalChartPeriod;
  }
  return TECHNICAL_CHART_DEFAULT_PERIOD;
}

/** Build an in-app href; only pass canonical stock code for cross-page entry. */
export function buildTechnicalChartHref(options: {
  stock: string;
  period?: TechnicalChartPeriod;
  days?: TechnicalChartDays | number;
  indicators?: string;
}): string {
  const params = new URLSearchParams();
  const stock = options.stock.trim();
  if (stock) {
    params.set('stock', stock);
  }
  if (options.period) {
    params.set('period', parseTechnicalChartPeriod(options.period));
  }
  if (options.days != null) {
    params.set('days', String(parseTechnicalChartDays(String(options.days))));
  }
  if (options.indicators) {
    params.set('indicators', parseTechnicalChartIndicators(options.indicators));
  }
  const query = params.toString();
  return query ? `/technical-chart?${query}` : '/technical-chart';
}

export const technicalChartApi = {
  get: async (params: GetTechnicalChartParams): Promise<TechnicalChartResponse> => {
    const {
      stockCode,
      period = TECHNICAL_CHART_DEFAULT_PERIOD,
      days = TECHNICAL_CHART_DEFAULT_DAYS,
      indicators,
      signal,
    } = params;
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/stocks/${encodeURIComponent(stockCode)}/technical-chart`,
      {
        params: {
          period: parseTechnicalChartPeriod(period),
          days,
          indicators: indicators || TECHNICAL_CHART_DEFAULT_INDICATORS,
        },
        signal,
      },
    );
    return toCamelCase<TechnicalChartResponse>(response.data);
  },
};
